pyo3-ffi/src/object.rs

```diff
@@ -91,6 +129,18 @@ pub unsafe fn Py_Is(x: *mut PyObject, y: *mut PyObject) -> c_int {
 }
 
 #[inline]
+#[cfg(Py_GIL_DISABLED)]
+pub unsafe fn Py_REFCNT(ob: *mut PyObject) -> Py_ssize_t {
+    let local = (AtomicU32::from((*ob).ob_ref_local)).load(Relaxed);
+    if local == _Py_IMMORTAL_REFCNT_LOCAL {
+        return _Py_IMMORTAL_REFCNT;
+    }
+    let shared = (AtomicIsize::from((*ob).ob_ref_shared)).load(Relaxed);
+    local as Py_ssize_t + Py_ssize_t::from(shared >> _Py_REF_SHARED_SHIFT)
```

Comment:
This is the only kind of hairy bit, since there's no function exposed in the API to get the refcount

This is my first time dealing with rust atomics. I have no idea if it's safe to create them inline and then do an atomic load like this.

Reply:
`AtomicU32::from((*ob).ob_ref_local)` creates a new atomic whose changes will not be reflected in the original `ob_ref_local`. So this is definitely wrong. `Atomic::from_ptr` would be more appropriate.

Anyway I would feel better if we just used the C functions to do refcounting, similar to https://github.com/PyO3/pyo3/issues/4311#issuecomment-2216740704. I'm sure @davidhewitt will have thoughts on this.

Reply:
Unfortunately there is no function for this in the API, only a static inline function.

Reply:
In https://github.com/PyO3/pyo3/issues/3357 we discussed maybe we should just remove this API.

Reply:
I think we still might need an implementation though if we need to expose it in the FFI for users of PyO3, e.g. as @davidhewitt implies in https://github.com/PyO3/pyo3/issues/3357#issuecomment-1661661193.

My understanding is that the plan for 3.14 is to either expose a function for this or to expose a function that provides similar functionality without needing to expose refcounts directly. Something like `Py_IsOnlyReference` for refcount==1 and otherwise everything else becomes an implementation detail. 

Unfortunately that doesn't help us for 3.13.

Reply:
I think that we probably do need to keep this API exposed (even after #4065 we use it for debugging and I think others might want to too). I guess that given it's a static inline function, let's just go down the `Atomic::from_ptr` route and hope that CPython can provide a C function in 3.14.